### ISSUE TITLE
SingStar games can't swap discs

### ISSUE THREAD

**Issue by fldef** [fldef/pcsx2-fldef#2]
Of note with the SingStar Pop demo:
* The game expects cdvd register 0x0b to reflect Media Changed status (see bug #1).
* The game expects cdvd status to be `CDVD_STATUS_SEEK` (0x12) following a disc swap. PSCX2 does not use that status.

**Comment by fldef** [comment]
@seta-san If you're interested in another try:

https://github.com/fldef/pcsx2-fldef/releases/tag/v1.6.0-swapfix-1

**Comment by seta-san** [comment]
> @seta-san If you're interested in another try:
> 
> https://github.com/fldef/pcsx2-fldef/releases/tag/v1.6.0-swapfix-1

this one is almost there. it works but only after swapping discs twice.

**Comment by fldef** [comment]
Great! If you get some time, could you post a video? I'll try to figure out what's missing from there.

**Comment by seta-san** [comment]
https://www.youtube.com/watch?v=k_WxFlom8Cc

**Comment by seta-san** [comment]
the process is you start by loading one of the discs. you get yourself to the swap disc screen. you swap your iso in pcsx2. when you do it sits there at the "Please Insert a SingStar Disc" screen. then you change the iso again in pcsx2 to a different one and that's when it finally works.

**Comment by fldef** [comment]
Thanks, that helps. Could you try swapping to non-SingStar games? I'm wondering if you get the same message on the first swap.

**Comment by seta-san** [comment]
First Swap give me the "Please insert a SingStar Disc" screen
on the second swap the screen very quickly clears followed by the same demand for a SingStar Disc.

**Comment by fldef** [comment]
Huh... I'll see if I can come up with anything.

**Comment by seta-san** [comment]
> Huh... I'll see if I can come up with anything.

there is something that is being set on the first swap attempt that isn't being acknowledged by the game/emulator until the second swap.

**Comment by fldef** [comment]
Could you try swapping first to "No disc" then to a different SingStar game?

**Comment by seta-san** [comment]
when i swap to "no disc" the console spams "*Unknown Mecacon Command param[0]=30" at me. but when i swap to a different singstar disc it works

**Comment by fldef** [comment]
One last try: https://github.com/fldef/pcsx2-fldef/releases/tag/v1.6.0-swapfix-2

This probably won't fix SingStar waiting on the first disc, but I think it will fix the Mecacon error with "no disc." 

One other thing to try: wait 10 seconds or so after your first swap, in case I have the state delays too long.

**Comment by seta-san** [comment]
You are going to do a pull request.. right?

**Comment by seta-san** [comment]
i'm really starting to think that the first time i try to change the iso it gives the signal to open the cdvd tray but when it loads it never actually gives the appropriate command to close it. when i go to open swap the disc a second time the system already thinks the disc tray is out so when the "eject button" is pressed again the system now sees the disc as being in the machine.

**Comment by fldef** [comment]
> You are going to do a pull request.. right?

I don't know. Regression testing in PCSX2 is impossible, only a few games are fixed here, swap problems have been passed over for years, these are low-level changes... The active maintainers probably won't want to take on the risk.

> i'm really starting to think that the first time i try to change the iso it gives the signal to open the cdvd tray but when it loads it never actually gives the appropriate command to close it.

Monster Rancher EVO works fine with the tray part, so it isn't exactly that. SingStar does at least one cdvd check that EVO doesn't (i.e., the `SEEK` status check). It's possible I'm missing others, like a timing problem or a hardware flag that I'm not aware of.

I might need to debug a SingStar swap to get further. Either way, thanks for your help working on this problem.

**Comment by seta-san** [comment]
Everything is there or else it wouldn't work on the second swap

**Comment by seta-san** [comment]
i found the solution. add that line and it'll work

 `static s32 cdvdCtrlTrayClose()
{
    DevCon.WriteLn(Color_Green, L"Close virtual disk tray");
    cdvd.Ready = CDVD_READY1;
    cdvd.TrayTimeout = TRAY_CLOSE_TRIGGER - 1; // Reset so it can't get closed twice by cdvdVsync()
    cdvd.MediaChanged = true;

    cdvdDetectDisk();
    GetCoreThread().ApplySettings(g_Conf->EmuOptions);
    **cdvd.Status = CDVD_STATUS_SEEK;**
    return 0; // needs to be 0 for success according to homebrew test "CDVD"
}`

**Comment by fldef** [comment]
That's great, nice work!

**Comment by fldef** [comment]
I have a SingStar game ordered so that I can test this better, and I'll see if I can get someone to test the Dynasty Warriors XL disc swapping, too. No worries, I haven't given up on anything. 👍 Just being cautious.

**Comment by fldef** [comment]
@seta-san I won't be able to test SingStar for myself until later this week. I have a new build if you're willing to test it. Thanks.

https://github.com/fldef/pcsx2-fldef/releases/tag/v1.6.0-swapfix-3

**Comment by seta-san** [comment]
Be happy to when I get home

**Comment by seta-san** [comment]
this latest build works most of the time.. but not every time

**Comment by fldef** [comment]
Thanks for the help. I'll have to test a SingStar game myself to figure it out.

**Comment by seta-san** [comment]
> Thanks for the help. I'll have to test a SingStar game myself to figure it out.

Keep in mind you don't have to swap the disc at the swap disc screen. . You need two singstar games to test

**Comment by fldef** [comment]
Swapping from SingStar ABBA (physical disc via Gigaherz plugin) to the SingStar demo (ISO) and back works every time for me.

Is there ever a scenario when swapping *always* works or *always* fails for you?
